### PR TITLE
Implement Polach Adhesion

### PR DESCRIPTION
**Pull request by peternewell**
Implements Polach adhesion physics to replace current adhesion physics.

This will allow future track based enhancements to be implemented

https://blueprints.launchpad.net/or/+spec/enhance-adhesion

### CODE REVIEW

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:955]: You have to comment this two lines when commenting the previous one

**Review by cesarBLG — COMMENTED** [review]
Okay, I have a couple of simple comments, which can be easily addressed. Formulas come from well known papers, and have been widely adopted. Code implementation is OK too.

I have some serious concerns about the suitability of the Polach model for OR, due to the amount of computational resources required. They won't prevent me from approving the PR once my other comments are addressed, because I feel like that's not my decision to make. Nevertheless, I want to leave a record of the problems here:

The fundamental problem with Polach model is that it predicts smaller wheel creeps than the model we were previously using. This is experimentally motivated, so in principle Polach model will provide more realistic wheel creeps. Therefore, it is desirable to have such a model in OR. Unfortunately, having lower wheel creeps comes at a cost. Wheel-rail dynamics are computed using a numerical RK4 integrator, but due to the properties of the adhesion formula (low wheel inertia, huge torque differences for small speed changes), numerical errors are high. This requires dividing the simulation in many substeps, until numerical error is kept low. The increased complexity of Polach formula, as well as the greater number of substeps required due to the smaller creeps, increase required CPU time. Possible solutions are either running at less substeps than required, running a noisy simulation that does not correspond to Polach model anymore, or running at full substeps impacting frame rates.

This basically forces people without high-end computers (including me) to switch to the simple adhesion model, when advanced adhesion worked well previously. I'd be completely in favour of using Polach formula if there were no drawbacks, but sadly that's not the case.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs]: Dead code removed (I hope, if I have understood your comment).

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSSteamLocomotive.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSSteamLocomotive.cs:5048]: I don't understand this typo. It is a temporary parameter only used in the following line, and was named slightly differently to differentiate it from the "real" parameter.

If you wish me to change it then can you explain in a bit more detail.

Thanks

File: Source/RunActivity/Viewer3D/Popups/HUDWindow.cs
Comment (peternewell) [reply Source/RunActivity/Viewer3D/Popups/HUDWindow.cs:479]: You maybe correct, but I am not sure how to call the base value that would be maintained up to date. 

Do you have any suggestions.

**Comment by peternewell** [comment]
> Okay, I have a couple of simple comments, which can be easily addressed. Formulas come from well known papers, and have been widely adopted. Code implementation is OK too.
> 
> I have some serious concerns about the suitability of the Polach model for OR, due to the amount of computational resources required. They won't prevent me from approving the PR once my other comments are addressed, because I feel like that's not my decision to make. Nevertheless, I want to leave a record of the problems here:
> 
> The fundamental problem with Polach model is that it predicts smaller wheel creeps than the model we were previously using. This is experimentally motivated, so in principle Polach model will provide more realistic wheel creeps. Therefore, it is desirable to have such a model in OR. Unfortunately, having lower wheel creeps comes at a cost. Wheel-rail dynamics are computed using a numerical RK4 integrator, but due to the properties of the adhesion formula (low wheel inertia, huge torque differences for small speed changes), numerical errors are high. This requires dividing the simulation in many substeps, until numerical error is kept low. The increased complexity of Polach formula, as well as the greater number of substeps required due to the smaller creeps, increase required CPU time. Possible solutions are either running at less substeps than required, running a noisy simulation that does not correspond to Polach model anymore, or running at full substeps impacting frame rates.
> 
> This basically forces people without high-end computers (including me) to switch to the simple adhesion model, when advanced adhesion worked well previously. I'd be completely in favour of using Polach formula if there were no drawbacks, but sadly that's not the case.

Thanks for offering to approve it despite your reservations.

I can understand your concerns, 

I believe that the potential performance issues are broader then just this PR, and I think that they should be raised with the ORMT at a project level.

Firstly if we look at the current minimum [PC requirements for OR](https://www.openrails.org/download/program/), we can see that they appear to be very low compared to some of our "competition", such as [Train Sim World](https://store.epicgames.com/en-US/p/train-sim-world-4) and [Just Trains](https://www.justtrains.net/product/train-sim-world-4-standard-edition). Interestingly the graphics RAM requirements are significantly higher for these products then the OR requirements.

So the first question that the ORMT needs to address is, "Do we maintain our current PC requirements, or should they be updated more in line with the competition?" If we maintain the current ones we will no doubt be restricting a lot of features due to PC performance issues, and we will need to consider an answer for the following question.

The second question then becomes, "Once we have set a standard, how do we test to make sure that OR is performing up to this standard?"

There also maybe other questions that need to be addressed.

Would you like me to raise this broader issue (or do you want to do it) with the ORMT?

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSSteamLocomotive.cs
Comment (cesarBLG) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSSteamLocomotive.cs:5048]: Ah, so using "postion" instead of "Position" is intentional? I didn't know that intentionally adding typos is a coding practice. I assume it is OK then, although I'd prefer using descriptive names for variables instead of adding typos. That's a personal preference though.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cesarBLG) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs]: Yes, I was referring to that commented part of code. Now that I see, also the if (count < 6 && count++ == 5) block is debug code which should be disabled. I'm not saying that it has to be deleted, but it should be enclosed in a #if #endif block.

File: Source/RunActivity/Viewer3D/Popups/HUDWindow.cs
Comment (cesarBLG) [reply Source/RunActivity/Viewer3D/Popups/HUDWindow.cs:479]: A good place may be MSTSLocomotiveViewer.PrepareFrame

**Comment by cesarBLG** [comment]
> Firstly if we look at the current minimum [PC requirements for OR](https://www.openrails.org/download/program/), we can see that they appear to be very low compared to some of our "competition", such as [Train Sim World](https://store.epicgames.com/en-US/p/train-sim-world-4) and [Just Trains](https://www.justtrains.net/product/train-sim-world-4-standard-edition). Interestingly the graphics RAM requirements are significantly higher for these products then the OR requirements.

Obviously we have much worse graphics than TSW and SimRail, so it's reasonable that we demand less powerful PCs.

> Would you like me to raise this broader issue (or do you want to do it) with the ORMT?

I think that would be a good idea.

For this specifical topic on adhesion, I believe the optimal thing to do is to include Simple Adhesion inside the Axle module, so the only difference between Simple and Advance adhesion is the wheel-rail contact, while using the same code for everything else.

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSSteamLocomotive.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSSteamLocomotive.cs]: Ok, I have renamed it for clarity.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs]: This code was not added by me, so I wasn't sure what it was.

I have set it up as debug code now.

File: Source/RunActivity/Viewer3D/Popups/HUDWindow.cs
Comment (peternewell) [reply Source/RunActivity/Viewer3D/Popups/HUDWindow.cs:479]: I am not certain as to how to render RenderProcess.FrameRate.SmoothedValue visible in MSTSLocomotiveViewer.PrepareFrame, so it will need to stay as it is unless I can get some code as an example of making it visible.

**Comment by peternewell** [comment]
> 
> Obviously we have much worse graphics than TSW and SimRail, so it's reasonable that we demand less powerful PCs.
> 
Having a better independent graphics card requirement would take some of the load off the PC rather then it trying to do some of the graphics processing.

> 
> I think that would be a good idea.
> 
I will raise it tomorrow.

>
> For this specifical topic on adhesion, I believe the optimal thing to do is to include Simple Adhesion inside the Axle module, so the only difference between Simple and Advance adhesion is the wheel-rail contact, while using the same code for everything else.
>
I would prefer to see what approach is identified by the ORMT first.

File: Source/RunActivity/Viewer3D/Popups/HUDWindow.cs
Comment (cesarBLG) [reply Source/RunActivity/Viewer3D/Popups/HUDWindow.cs:479]: But this only works if the HUD is open, which is not ideal. Why don't you use elapsedClockSeconds, which is directly obtained at the Axle.Update function, and smooth that yourself if required?

Comment (peternewell) [reply Source/RunActivity/Viewer3D/Popups/HUDWindow.cs:479]: I would prefer to use the FrameRate values, as this is the parameter which is being impacted, otherwise we would need to find the correct correlation between the FrameRate and various values of elapsedClockSeconds.

I suspect that most people will have the HuD open (even if it is only the main HuD), so this should be ok.

I would like to have this approved now and then we will see what other "negative" or "positive" feedback that we get in the next stage.

As a background task, I will see if I can find a way to read the FrameRate "continuously".

Comment (cesarBLG) [reply Source/RunActivity/Viewer3D/Popups/HUDWindow.cs:479]: > I suspect that most people will have the HuD open (even if it is only the main HuD), so this should be ok.
That's a huge assumption, and I don't think it is true. Anyway, limiting adhesion to work only when the HUD is opened is a mistake. I'm sorry, but I cannot approve it until this is fixed. We don't have to be in a hurry to merge PR which contains serious defects as this one, but we should take some time to fix them instead of including bad code into OR.

> I would prefer to use the FrameRate values, as this is the parameter which is being impacted, otherwise we would need to find the correct correlation between the FrameRate and various values of elapsedClockSeconds.
Your code is basically saying: "if frame rate is low, I do not care about you, I'll just give lots of numerical errors to force you to switch to simple adhesion" I don't feel like this is the right approach, but since there is no agreement on this, I won't make further comments about it.

Re elapsedClockSeconds and FPS, when simulation time is at 100%, FPS=1/elapsedClockSeconds, so you could possibly use that.

**Comment by cjakeman** [comment]
Hi Peter and Cesar,

I've sent my thoughts to James and not had a reply yet. I'll repeat them here and see if he agrees with me.

1. Firstly, a big "thank you" for your work in improving our adhesion model. I expect it will make quite a difference in the users' experience.

2. On the subject of the recommended system for Open Rails, we are introducing a Notification mechanism into Menu.exe. As well as giving OR (and also content providers) a chance to notify users about changes and improvements, it will also gather system specs from a sample of willing users. That will give us sound data on which to base upgrade decisions (such as moving off Windows 7 in order to use glTD).

It would be neat if RunActivity can also use that mechanism to notify the user about issues that might otherwise be overlooked in the log file. For example: 
"OR is not using advanced adhesion as this PC lacks performance. [More details]". 

3. In an earlier message, James agreed with Cesar that the best measure for adjusting the calculation is elapsedSeconds. Basically, the Simulator must not access properties belonging to the Viewer, such as frame rate.

4. Regarding the Polach algorithm, we do not want to take functionality away from users with low-spec PCs (e.g. Cesar). Is there any problem in OR using Polach initially and automatically downgrading to the current algorithm if elapsedSeconds rises above a threshold value? Perhaps with a warning in the log file?

Best wishes,
Chris

**Comment by peternewell** [comment]
Hi Chris,

Thanks for the response.

> 
> It would be neat if RunActivity can also use that mechanism to notify the user about issues that might otherwise be overlooked in the log file. For example: "OR is not using advanced adhesion as this PC lacks performance. [More details]".
> 
Ok, I will add something along this line, however as the FPS seems to be impacted, then my preference would be to provide a reading of it in the error message for debugging. 

>
>     3. In an earlier message, James agreed with Cesar that the best measure for adjusting the calculation is elapsedSeconds. Basically, the Simulator must not access properties belonging to the Viewer, such as frame rate.
> 
Again my preference would be to monitor the impacted parameter, ie FrameRate, as I am wondering whether a good graphics card mitigates the performance impact, and thus elapsedSeconds doesn't give a correlation to FrameRate. However I will be guided by your expertise.

> 
>     4. Regarding the Polach algorithm, we do not want to take functionality away from users with low-spec PCs (e.g. Cesar). Is there any problem in OR using Polach initially and automatically downgrading to the current algorithm if elapsedSeconds rises above a threshold value? Perhaps with a warning in the log file?
> 
Ok, I will have a go at your suggestion.

I imagine that we need to determine the threshold limits as follows:

Lower Limit (switch to existing adhesion model) - my thoughts in FPS = 55

Upper Limit (switch to Polach adhesion model) - my thoughts in FPS = 60

Do you have suggested values for the above limits?

**Comment by peternewell** [comment]
I have now uploaded a patch to switch between the Pacha and Polach adhesion models.

At the moment it still uses the FrameRate to switch. It switches down @ 55 and up at @ 59 (60 fps seems to be the maximum display value). I am still not convinced that timeSpan is a close measure of video performance, but I am open to suggestions if you provide some values for me.

I have set up the WheelAdhesion value to read 99% when in the Pacha model.

Currently the user cannot select which model that they can use. I also think that we need to be careful adding a lot of variations like this as it makes OR more complex and harder to debug.

Anyway hopefully this will suit all parties and allow the PR to be approved.

Thanks

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cjakeman) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:432]:
```suggestion
        bool UsePolachAdhesion = false;
```

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cjakeman) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:861]:
```suggestion
        /// The folloi=wing section attempts to calculate the optimal substep limit. This is a trade off between the accuracy of the slips calculations and the CPU load which impacts the screen FPS
```
        /// The following section attempts to calculate the optimal substep limit. This is a trade off between the accuracy of the slips calculations and the CPU load which impacts the screen FPS

**Comment by cjakeman** [comment]
Hi Peter,

> I have now uploaded a patch to switch between the Pacha and Polach adhesion models.
> 
> At the moment it still uses the FrameRate to switch. It switches down @ 55 and up at @ 59 (60 fps seems to be the maximum display value). I am still not convinced that timeSpan is a close measure of video performance, but I am open to suggestions if you provide some values for me.

I'm pleased that you were able to incorporate both schemes. I had imagined that switching down to Pacha would persist to the end of the game, but I can see that your 2-way switch with hysteresis might be even better. It's good to see all the comments explaining what's happening and why.

> Currently the user cannot select which model that they can use. I also think that we need to be careful adding a lot of variations like this as it makes OR more complex and harder to debug.

I'm sure you are right to be cautious - I believe it's known as a "combinatorial explosion". On the forum you talk about simulating a low-spec PC so we can test for a range of user systems. I'll ask James about that.

> Anyway hopefully this will suit all parties and allow the PR to be approved.

Well we can't let it through while it uses framerates directly as the Simulator must not spy on the Viewer properties. I will get advice on the ElapsedSeconds that are most appropriate for your switch points.

Again, my thanks for all this work,

Chris

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs]: Spelling errors corrected. Thanks

**Review by cjakeman — COMMENTED** [review]
Hi Peter,

Doesn't this enhancemet warrant some changes to the Manual?

Also I don't think the Manual says anything at all about the impact of adhesion on pulling a heavy train. Perhaps we could get someone else to write that section?

Best wishes,

Chris

**Comment by peternewell** [comment]
Hi Chris,

> 
> Doesn't this enhancemet warrant some changes to the Manual?
> 
> Also I don't think the Manual says anything at all about the impact of adhesion on pulling a heavy train. Perhaps we could get someone else to write that section?
> 
> Best wishes,
> 
I have added some notes to the manual in regard to the algorithm switching that OR does for lower performance PCs.

If you have anybody in mind to write the "impact of adhesion" piece then that would be fine.

Thanks

Peter

**Comment by cesarBLG** [comment]
> Also I don't think the Manual says anything at all about the impact of adhesion on pulling a heavy train. Perhaps we could get someone else to write that section?

Hi Chris,

Wheel-rail adhesion is a local interaction, and it is independent of train weight. The difference comes from the fact that when pulling a heavy train, you will have to set a higher throttle setting to be able to accelerate. Higher throttle settings are more likely to produce wheelslip. However, for a given throttle setting the slip probability is the same irregardless of the pulled weight.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:830]: I would prefer if the differences between models are kept in SlipCharacteristics(), so the rest of the functions is platform-agnostic. I offer to do that in a further PR if you prefer. Only a couple of lines have to be changed.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:987]: This can also be simplified to make it model-agnostic

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:994]: This was debug info that isn't used anymore. We can remove it.

**Review by cesarBLG — COMMENTED** [review]
Hi Peter,
I have a style suggestion on how to switch between models without having lots of if (usePolachAdhesion) clauses. I can fix it myself when this gets merged, so you can switch to a different topic.

**Comment by cjakeman** [comment]
Hi Peter,

Thanks for adding a piece for the manual.

I asked James about using the Frame Rate and it seems I was not understanding it fully. The Updater thread (where the Simulator updates) and the Render thread work in strict sequence so there is exactly 1 simulator cycle for each frame. So the relationship betwwen frame rate and elapsed seconds is fixed (as in this code snippet from RenderProcess.cs):

```
        public void ComputeFPS(float elapsedRealTime)
        {
            if (elapsedRealTime < 0.001)
                return;
            FrameRate.Update(elapsedRealTime, 1f / elapsedRealTime);
            FrameTime.Update(elapsedRealTime, elapsedRealTime);
        }
```

I hope that makes things clearer.

> At the moment it still uses the FrameRate to switch. It switches down @ 55 and up at @ 59 (60 fps seems to be the maximum display value). 

We think that you're setting the bar rather high for Polach Adhesion and switching at 30 and 40 fps might be better.

The real test is how well this suits Cesar and others with low-spec PCs. Would you use your Polach forum thread to ask for feedback on exactly this issue? If that proves to be OK, then we can approve your PR.

Best wishes,

File: Source/Documentation/Manual/physics.rst
Comment (peternewell) [reply Source/Documentation/Manual/physics.rst]: Spelling error corrected.

**Comment by peternewell** [comment]
Hi Chris,
> 
> The real test is how well this suits Cesar and others with low-spec PCs. Would you use your Polach forum thread to ask for feedback on exactly this issue? If that proves to be OK and SetFrameRate() is removed, then your PR can be approved and merged.
> 
I have made the changes that you have requested.

Hopefully given the fact that there were not many people raising issues (probably won't impact many users until it goes to Testing) I am hoping that you can approve it within the next couple of days so that I can move onto some changes that I would like to add and merging this will help.

I am happy to address any issues addressed longer term if required.

Thanks

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:960]: We have to add an additional check so that we don't switch between models when simulation is paused (timeSpan=0). Also, when entering dense areas, frame rate will drop and will cause a non-smooth transition between models. That's not good I think.

File: Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs
Comment (cesarBLG) [inline Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:976]: Why are we setting this to 99% instead of using the same logic?

Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:976]: >
> Why are we setting this to 99% instead of using the same logic?
>
Both of these values were not displayed in the Pacha model. Wheel Adhesion could be displayed, but MaxWheelAdhesion was not calculated for the Pacha model.

Perhaps the more elegant way would be to change the HuD so that these lines are removed when switched to the Pacha mode.

Comment (cesarBLG) [reply Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/PowerTransmissions/Axle.cs:976]: But you can calculate MaxWheelAdhesion for the Pacha model too. It may be good having it for debugging. Never mind, I'll refactor this later to make almost everything in axle.cs model independent.

**Comment by peternewell** [comment]
> Hi Peter, I see some potential issues that I would prefer to be addressed. Maybe I can approve it as is, and then we fix the problems in the future? I think none of us really knows what's happening with the substeps and its relationship with FPS, which is why we are trying some complex workarounds.
> 
> For example, in my case I typically get low FPS, but I don't know if the bottleneck is in the Update process or in the Render process. If the bottleneck is caused by rendering, this means that we can safely increase the number of substeps, and the FPS won't be impacted (the Update thread is suspended until restarted by the Render process, so there is room for extra CPU usage). On the other hand, if low FPS are caused by the Update process, then increasing the number of substeps will degrade performance. Since difference in CPU performance is quite similar for all computers, I would expect the bottleneck to be in the graphics side.
> 
> I write this because, with our current FPS-based approach, there will be jumps in wheel speed when FPS varies, and this will impact user experience. If we are sure that using 100 substeps does not impact frame rates (i.e. we have a rendering bottleneck), then we could use Polach for all systems.
> 
> All this discussion is beyond my knowledge, so I may be wrong.
>
It has always been my concern that the GPU card could be masking the problem.

James (via info from Chris) suggests that the Render and Update processes shouldn't be causing each other issues.

So the question becomes how do we confirm it one way or another. 

I have a almost new PC and I run at 60fps all the time so I am not a good tester. Are you able to put any debug code into a test branch and see if you can confirm it one way or another?

As you suggest, I think that we are getting to the stage of approving it, and then addressing any potential issues identified, if they are raised.

**Review by cjakeman — APPROVED** [review]
Hi Peter,

Thanks for making these changes.

Approved and merged.

Chris

**Comment by peternewell** [comment]
Thanks Chris

**Pull request merged**
